src/devices_models/devices/thermal_generation.jl

```diff
@@ -62,189 +63,128 @@ get_variable_binary(::Union{ColdStartVariable, WarmStartVariable, HotStartVariab
+    [OnVariable()]
 end
 
-"""
-This function adds the active power limits of generators when there are CommitmentVariables
-"""
-function DeviceRangeConstraintSpec(
-    ::Type{<:ActivePowerVariableLimitsConstraint},
-    ::Type{ActivePowerVariable},
-    ::Type{T},
-    ::Type{<:AbstractThermalUnitCommitment},
-    ::Type{<:PM.AbstractPowerModel},
-    feedforward::Nothing,
-    use_parameters::Bool,
-) where {T <: PSY.ThermalGen}
-    return DeviceRangeConstraintSpec(;
-        range_constraint_spec = RangeConstraintSpec(;
-            constraint_type = ActivePowerVariableLimitsConstraint(),
-            variable_type = ActivePowerVariable(),
-            bin_variable_types = [OnVariable()],
-            limits_func = x -> PSY.get_active_power_limits(x),
-            constraint_func = device_semicontinuousrange!,
-            constraint_struct = DeviceRangeConstraintInfo,
-            component_type = T,
-        ),
+
+function add_constraints!(
+    container::OptimizationContainer,
+    T::Type{<:PowerVariableLimitsConstraint},
+    U::Type{<:VariableType},
+    devices::IS.FlattenIteratorWrapper{V},
+    model::DeviceModel{V, W},
+    X::Type{<:PM.AbstractPowerModel},
+    feedforward::Union{Nothing, AbstractAffectFeedForward}
+) where {
+    V <: PSY.ThermalGen,
+    W <: ThermalCompactDispatch,
+}
+    add_range_constraints!(container, T, U, devices, model, X, feedforward)
+end
+
+function get_limits(device, ::Type{PowerAboveMinimumVariable}, ::Type{<:ThermalCompactDispatch})
+    (
+        min = 0.0,
+        max = PSY.get_active_power_limits(device).max -
+                PSY.get_active_power_limits(device).min,
     )
 end
 
-"""
-This function adds the active power limits of generators when there are no CommitmentVariables
-"""
-function DeviceRangeConstraintSpec(
-    ::Type{<:ActivePowerVariableLimitsConstraint},
-    ::Type{ActivePowerVariable},
-    ::Type{T},
-    ::Type{<:ThermalDispatchNoMin},
-    ::Type{<:PM.AbstractPowerModel},
-    feedforward::Nothing,
-    use_parameters::Bool,
-) where {T <: PSY.ThermalGen}
-    return DeviceRangeConstraintSpec(;
-        range_constraint_spec = RangeConstraintSpec(;
-            constraint_type = ActivePowerVariableLimitsConstraint(),
-            variable_type = ActivePowerVariable(),
-            limits_func = x -> (min = 0.0, max = PSY.get_active_power_limits(x).max),
-            constraint_func = device_range!,
-            constraint_struct = DeviceRangeConstraintInfo,
-            component_type = T,
-        ),
-        custom_optimization_container_func = custom_active_power_constraints!,
+
+# active power limits of generators when there are no CommitmentVariables
+# TODO: handle custom active power constraints for optimization
+function get_limits(device, ::Type{ActivePowerVariableLimitsConstraint}, ::Type{<:ThermalDispatchNoMin}) #  -> Union{Nothing, NamedTuple{(:startup, :shutdown), Tuple{Float64, Float64}}}
```

Comment:
If this comment is resolved can you remove it?

Reply:
I don't think this one is resolved? See the following:

https://github.com/NREL-SIIP/PowerSimulations.jl/blob/d757cc33435ede264cf873e1d2eec97ae0d3330c/src/devices_models/devices/thermal_generation.jl#L172

Reply:
We don't need custom power constraints now that everything is implemented in `add_constraint` and not in the spec. If a user needs a custom implementation then it can add the JuMP code directly into the `add_constraint` code.